processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -149,4 +155,178 @@ public void remove()
         }
     );
   }
+
+
+
+  public static Filter convertToCNFFromQueryContext(Query query, Filter filter)
```

Comment:
It doesn't appear to take that long; with this benchmark that converts a somewhat complex filter to CNF:

```
  @Setup
  public void setup()
  {
    DimFilter dimFilter1 = new OrDimFilter(Arrays.<DimFilter>asList(
        new SelectorDimFilter("dimA", "199", null),
        new AndDimFilter(Arrays.<DimFilter>asList(
            new NoBitmapSelectorDimFilter("dimF", "Corundum", null),
            new SelectorDimFilter("dimE", "bar", null)
        )
        ))
    );
    DimFilter dimFilter2 = new OrDimFilter(Arrays.<DimFilter>asList(
        new SelectorDimFilter("dimA", "299", null),
        new SelectorDimFilter("dimA", "399", null),
        new AndDimFilter(Arrays.<DimFilter>asList(
            new NoBitmapSelectorDimFilter("dimF", "Xylophone", null),
            new SelectorDimFilter("dimE", "foo", null)
        )
        ))
    );
    DimFilter dimFilter3 = new OrDimFilter(Arrays.<DimFilter>asList(
        dimFilter1,
        dimFilter2,
        new AndDimFilter(Arrays.<DimFilter>asList(
            new NoBitmapSelectorDimFilter("dimF", "Orange", null),
            new SelectorDimFilter("dim2", "World", null)
        )
        ))
    );

    filter = dimFilter3.toFilter();
  }

  @Benchmark
  @BenchmarkMode(Mode.AverageTime)
  @OutputTimeUnit(TimeUnit.MICROSECONDS)
  public void convertFilterCNF(Blackhole blackhole) throws Exception
  {
    Filter filter2 = Filters.convertToCNF(filter);
    blackhole.consume(filter2);
  }
```

takes ~4.8us for the conversion:

```
Benchmark                                                  Mode  Cnt  Score   Error  Units
FilterPartitionBenchmark.readComplexOrFilterCNF            avgt   25  4.808 ± 0.121  us/op
```